content/articles/2019-08-callr-task-q.Rmd

```diff
@@ -0,0 +1,586 @@
+      pr <- processx::poll(conns, as_ms(timeout))
+      private$tasks$state[topoll][pr == "ready"] <- "ready"
+      private$schedule()
+      private$tasks$id[private$tasks$state == "done"]
+    }
+```
+
+We only need to check on tasks that are _running_. `poll()` uses the
+`processx::poll()` function that can wait on several `callr::r_session`s
+at once. More precisely, I extract the _poll connections_ of the
+`r_session` objects and call `processx::poll()` on these. An `r_session`
+may have multiple pollable connections, one for its standard output
+stream, one for its standard error stream. These are not used by default
+in `r_session`s, and I only want to check on the poll connection, which
+signals if the R session has finished with the computation (or encountered
+an error while working on it). `processx::poll()` returns a list of
+character vectors, one entry for each (_running_) task. This is `"ready"`
+if the session is ready with the task. (Or it is `"silent"` if it is not
+ready, or `"timeout"` if the time limit expired and no workers are ready.)
+
+All tasks that returned `"ready"` are indeed set to the _ready_ state.
+After this `poll()` calls `schedule()` to read out the results of the
+_ready_ tasks and reassign their workers to _waiting_ ones.
+
+`poll()` returns the ids of all tasks that are _done_.
+
+This version of `poll()` has a small issue when the R sessions are starting
+up: it might return without any results, before the specified timeout
+value is over. At startup the idle tasks are _running,_ and they are
+polled by `processx::poll()`. If any of the R sessions start up before the
+timeout is over, `processx::poll()` returns with `"ready"` for them.
+But `schedule()` cannot mark these tasks as _done_, because they are idle
+tasks, they'll be _waiting_, and with no task _done_, `poll()`
+will return an empty vector. This is problematic, because `poll()`
+promises to either wait until the specified timeout _or_ return a task
+that is _done_. So we need to wrap the simplified `poll()` into a loop, and
+keep calling `processx::poll()` until either the timeout
+expires or a task is _done_. The final `poll()` looks like this:
+
+```r
+    poll = function(timeout = 0) {
+      limit <- Sys.time() + timeout
+      as_ms <- function(x)
+        if (x==Inf) -1 else as.integer(as.double(x, "secs") * 1000)
+      repeat{
+        topoll <- which(private$tasks$state == "running")
+        conns <- lapply(
+          private$tasks$worker[topoll],
+          function(x) x$get_poll_connection())
+        pr <- processx::poll(conns, as_ms(timeout))
+        private$tasks$state[topoll][pr == "ready"] <- "ready"
+        private$schedule()
+        ret <- private$tasks$id[private$tasks$state == "done"]
+        if (is.finite(timeout)) timeout <- limit - Sys.time()
+        if (length(ret) || timeout < 0) break;
+      }
+      ret
+    }
+```
+
+Unfortunately `processx::poll()` expects the timeout as an integer in
```

Comment:
As above, I think you should just pass number of milliseconds directly in order to focus more on the interesting details.